File: exercise/java/day24/src/main/java/domain/core/EventSourcedAggregate.java
package domain.core;

import io.vavr.collection.List;
import io.vavr.collection.Seq;
import lombok.Getter;

import java.time.LocalDateTime;
import java.util.UUID;
import java.util.function.Consumer;
import java.util.function.Supplier;

public abstract class EventSourcedAggregate implements Aggregate {
    private List<Event> uncommittedEvents = List.empty();
    private final RegisteredRoutes registeredRoutes = new RegisteredRoutes();
    private final Supplier<LocalDateTime> timeProvider;

    @Getter
    private UUID id;
    @Getter
    private int version;

    protected EventSourcedAggregate(Supplier<LocalDateTime> timeProvider) {
        this.timeProvider = timeProvider;
        registerRoutes();
    }

    protected void setId(UUID id) {
        this.id = id;
    }

    protected abstract void registerRoutes();

    @Override
    public void applyEvent(Event event) {
        registeredRoutes.dispatch(event);
        version++;
    }

    protected <E extends Event> void registerEventRoute(Class<E> eventType, Consumer<E> apply) {
        registeredRoutes.register(eventType.getName(), event -> apply.accept(eventType.cast(event)));
    }

    @Override
    public Seq<Event> getUncommittedEvents() {
        return uncommittedEvents;
    }

    @Override
    public void clearUncommittedEvents() {
        uncommittedEvents = List.empty();
    }

    protected void raiseEvent(Event event) {
        applyEvent(event);
        uncommittedEvents = uncommittedEvents.append(event);
    }

    @Override
    public int hashCode() {
        return id.hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        return this == obj || (obj instanceof Aggregate other && id.equals(other.getId()));
    }

    protected LocalDateTime time() {
        return timeProvider.get();
    }
}

File: exercise/kotlin/day24/src/main/kotlin/domain/core/EventSourcedAggregate.kt
package domain.core

import java.time.LocalDateTime
import java.util.*

abstract class EventSourcedAggregate(private val timeProvider: () -> LocalDateTime) : Aggregate {
    private var uncommittedEvents: List<Event> = emptyList()
    private val registeredRoutes: RegisteredRoutes = RegisteredRoutes()
    override var id: UUID = UUID.randomUUID()
    override var version: Int = 0

    init {
        this.registerRoutes()
    }

    protected abstract fun registerRoutes()
    override fun applyEvent(event: Event) {
        registeredRoutes.dispatch(event)
        version++
    }

    protected fun <E : Event> registerEventRoute(eventType: Class<E>, func: (event: E) -> Unit) =
        registeredRoutes.register(eventType.name) { event -> func(eventType.cast(event)!!) }

    override fun getUncommittedEvents(): List<Event> = uncommittedEvents
    override fun clearUncommittedEvents() {
        uncommittedEvents = emptyList()
    }

    protected fun raiseEvent(event: Event) {
        applyEvent(event)
        uncommittedEvents = uncommittedEvents + event
    }

    override fun hashCode(): Int = id.hashCode()
    override fun equals(other: Any?): Boolean = this === other || (other is Aggregate && id == other.id)

    protected fun time(): LocalDateTime = timeProvider()
}
